components/mosquitto/examples/broker/main/broker.c

```diff
@@ -0,0 +1,92 @@
+#include "esp_log.h"
+#include "mosq_broker.h"
+#include "protocol_examples_common.h"
+
+const static char *TAG = "mqtt_broker";
+
+#if CONFIG_EXAMPLE_BROKER_RUN_LOCAL_MQTT_CLIENT
+static void mqtt_event_handler(void *handler_args, esp_event_base_t base, int32_t event_id, void *event_data)
+{
+    ESP_LOGD(TAG, "Event dispatched from event loop base=%s, event_id=%" PRIi32 "", base, event_id);
+    esp_mqtt_event_handle_t event = event_data;
+    esp_mqtt_client_handle_t client = event->client;
+    int msg_id;
+    switch ((esp_mqtt_event_id_t)event_id) {
+    case MQTT_EVENT_CONNECTED:
+        ESP_LOGI(TAG, "MQTT_EVENT_CONNECTED");
+        msg_id = esp_mqtt_client_subscribe(client, "/topic/qos0", 0);
+        ESP_LOGI(TAG, "sent subscribe successful, msg_id=%d", msg_id);
+        break;
+    case MQTT_EVENT_DISCONNECTED:
+        ESP_LOGI(TAG, "MQTT_EVENT_DISCONNECTED");
+        break;
+
+    case MQTT_EVENT_SUBSCRIBED:
+        ESP_LOGI(TAG, "MQTT_EVENT_SUBSCRIBED, msg_id=%d", event->msg_id);
+        msg_id = esp_mqtt_client_publish(client, "/topic/qos0", "data", 0, 0, 0);
+        ESP_LOGI(TAG, "sent publish successful, msg_id=%d", msg_id);
+        break;
+    case MQTT_EVENT_UNSUBSCRIBED:
+        ESP_LOGI(TAG, "MQTT_EVENT_UNSUBSCRIBED, msg_id=%d", event->msg_id);
+        break;
+    case MQTT_EVENT_PUBLISHED:
+        ESP_LOGI(TAG, "MQTT_EVENT_PUBLISHED, msg_id=%d", event->msg_id);
+        break;
+    case MQTT_EVENT_DATA:
+        ESP_LOGI(TAG, "MQTT_EVENT_DATA");
+        ESP_LOGI(TAG, "TOPIC=%.*s", event->topic_len, event->topic);
+        ESP_LOGI(TAG, "DATA=%.*s", event->data_len, event->data);
+
+        break;
+    case MQTT_EVENT_ERROR:
+        ESP_LOGI(TAG, "MQTT_EVENT_ERROR");
+        break;
+    default:
+        ESP_LOGI(TAG, "Other event id:%d", event->event_id);
+        break;
+    }
+}
+
+
+static void mqtt_app_start(void *ctx)
+{
+    struct mosq_broker_config *config = ctx;
+
+    vTaskDelay(pdMS_TO_TICKS(1000));    // wait a second for the broker to start
+    esp_mqtt_client_config_t mqtt_cfg = {
+        .broker.address.hostname = "127.0.0.1",
+        .broker.address.transport = MQTT_TRANSPORT_OVER_TCP,    // we support only TCP transport now
+        .broker.address.port = config->port,
+    };
+    esp_mqtt_client_handle_t client = esp_mqtt_client_init(&mqtt_cfg);
+    esp_mqtt_client_register_event(client, ESP_EVENT_ANY_ID, mqtt_event_handler, NULL);
+    esp_mqtt_client_start(client);
+    vTaskDelete(NULL);
+}
+#endif // CONFIG_EXAMPLE_BROKER_RUN_LOCAL_MQTT_CLIENT
+
+void app_main(void)
+{
+    ESP_ERROR_CHECK(nvs_flash_init());
+    ESP_ERROR_CHECK(esp_netif_init());
+    ESP_ERROR_CHECK(esp_event_loop_create_default());
+    ESP_ERROR_CHECK(example_connect());
+
+    struct mosq_broker_config config = { .host = CONFIG_EXAMPLE_BROKER_HOST, .port = CONFIG_EXAMPLE_BROKER_PORT };
+
+#if CONFIG_EXAMPLE_BROKER_RUN_LOCAL_MQTT_CLIENT
+    xTaskCreate(mqtt_app_start, "mqtt_client", 4096, &config, 4, NULL);
```

Comment:
Instead of spawning a task, have you considered adding the task delay on the before connect event handler?

Reply:
Good point, thanks! Removed the extra task and using before connect event to stall the mqtt task,